include/ensmallen_bits/moead/moead_impl.hpp

```diff
@@ -0,0 +1,418 @@
+                    const double crossoverProb,
+                    const double mutationProb,
+                    const double mutationStrength,
+                    const size_t neighbourhoodSize,
+                    const double distributionIndex,
+                    const double lowerBound,
+                    const double upperBound) :
+    populationSize(populationSize),
+    numGeneration(numGeneration),
+    crossoverProb(crossoverProb),
+    mutationProb(mutationProb),
+    mutationStrength(mutationStrength),
+    neighbourhoodSize(neighbourhoodSize),
+    distributionIndex(distributionIndex),
+    lowerBound(lowerBound * arma::ones(1, 1)),
+    upperBound(upperBound * arma::ones(1, 1)),
+    numObjectives(0)
+  { /* Nothing to do here. */ }
+
+//! Optimize the function.
+template<typename MatType,
+         typename... ArbitraryFunctionType,
+         typename... CallbackTypes>
+typename MatType::elem_type MOEAD::Optimize(std::tuple<ArbitraryFunctionType...>& objectives,
+                                            MatType& iterate,
+                                            CallbackTypes&&... callbacks)
+{
+  // Check if lower bound is a vector of a single dimension.
+  if (lowerBound.n_rows == 1)
+    lowerBound = lowerBound(0, 0) * arma::ones(iterate.n_rows, iterate.n_cols);
+
+  // Check if lower bound is a vector of a single dimension.
+  if (upperBound.n_rows == 1)
+    upperBound = upperBound(0, 0) * arma::ones(iterate.n_rows, iterate.n_cols);
+  // Convenience typedefs.
+  typedef typename MatType::elem_type ElemType;
+
+  if(lowerBound.size() != upperBound.size())
+  {
+    throw std::logic_error("MOEAD::Optimize(): size of lowerBound and upperBound "
+        "must be equal.");
+  }
+  if(lowerBound.size() != iterate.n_elem)
+  {
+    throw std::logic_error("MOEAD::Optimize(): there should be a lower bound and "
+        "an upper bound for each variable in the initial point.");
+  }
+  // Number of objective functions. Represented by m in the paper.
+  numObjectives = sizeof...(ArbitraryFunctionType);
+
+  // Controls early termination of the optimization process.
+  bool terminate = false;
+
+  // 1.1 The external population, non-dominated solutions.
+  std::vector<MatType> externalPopulation;
+  std::vector<arma::vec> externalPopulationFValue;
+
+  arma::Col<size_t> shuffle;
+  // Weight vectors, where each one of them represents a decomposition.
+  arma::Mat<ElemType> weights(numObjectives, populationSize, arma::fill::randu);
+
+  // 1.2 Storing the indices of nearest neighbours of each weight vector.
+  arma::Mat<size_t> weightNeighbourIndices(populationSize, neighbourhoodSize);
+  for (size_t i = 0; i < populationSize; i++)
+  {
+    // To temporarily store the distance between weights(i) and each other weights.
+    arma::vec distances(populationSize);
+    for (size_t j = 0; j < populationSize; j++)
+      distances(j) = std::sqrt(arma::accu(arma::pow(weights.col(i)-weights.each_col(), 2)));
```

Comment:
```suggestion
      distances(j) = std::sqrt(arma::accu(arma::pow(weights.col(i) - weights.each_col(), 2)));
```

Minor style fix, missing space between minus operator.